### PR TITLE
Process Validator Deposits

### LINKED ISSUE #1102: Implement Process Block Deposits
Hi all,

This is a tracking issue for per-block processing to also process deposits coming from PoW receipt roots in the PoW chain. Given this is an active area of research and refactor in the [specification](https://github.com/ethereum/eth2.0-specs/blob/master/specs/core/0_beacon-chain.md#deposits-1), this will be labeled as blocked by research until it is safer to implement.

### PR DESCRIPTION
**Pull request by rauljordan**
This resolves #1102

---

# Description

This PR implements verification and processing a block's deposits using outlined above using the official spec's deposit processing section at commit hash [6b84dae09e38808f95eb299c97f945be991535bb](https://github.com/ethereum/eth2.0-specs/blob/6b84dae09e38808f95eb299c97f945be991535bb/specs/core/0_beacon-chain.md).

## Requirements for Merge
- [x] Process Deposit function
- [x] ProcessBlockValidatorDeposits block operation logic
- [x] Proto field updates to match spec for deposits
- [x] Decoding deposit data
- [x] Verify merkle root function
- [x] Tests

### CODE REVIEW

**Comment by terenc3t** [comment]
Will review tonight. In the meantime, lint is failing 
` block_operations.go:636:21: error strings should not be capitalized or end with punctuation or a newline`

File: beacon-chain/core/blocks/block.go [hunk 1]
```diff
@@ -131,3 +134,30 @@ func DomainVersion(data *pb.ForkData, slot uint64, domainType uint64) uint64 {
...
+	depositInput *pb.DepositInput,
+	depositValue uint64,
+	depositTimestamp int64,
+) ([]byte, error) {
+	wBuf := new(bytes.Buffer)
+	if err := ssz.Encode(wBuf, depositInput); err != nil {
+		return nil, fmt.Errorf("failed to encode deposit input: %v", err)
+	}
+	encodedInput := wBuf.Bytes()
+	depositData := []byte{}
```
Comment (nisdas) [inline beacon-chain/core/blocks/block.go]: could we predefine the size of this ```depositData := make([]byte,0,16+len(encodedInput))``` ?

File: beacon-chain/core/blocks/block_operations.go [hunk 2]
```diff
@@ -524,6 +527,136 @@ func verifyAttestation(beaconState *pb.BeaconState, att *pb.Attestation) error {
...
+		return nil, fmt.Errorf(
+			"number of deposits (%d) exceeds allowed threshold of %d",
+			len(deposits),
+			params.BeaconConfig().MaxDeposits,
+		)
+	}
+	var err error
+	var depositInput *pb.DepositInput
+	for idx, deposit := range deposits {
+		depositData := deposit.GetDepositData()
```
Comment (nisdas) [inline beacon-chain/core/blocks/block_operations.go]: Change this to `deposit.DepositData`

File: beacon-chain/core/blocks/block_operations.go [hunk 3]
```diff
@@ -524,6 +527,136 @@ func verifyAttestation(beaconState *pb.BeaconState, att *pb.Attestation) error {
...
+	}
+	return depositInput, nil
+}
+
+func verifyDeposit(beaconState *pb.BeaconState, deposit *pb.Deposit) error {
+	depositData := deposit.GetDepositData()
+	// Verify Merkle proof of deposit and PoW receipt trie root.
+	var receiptRoot [32]byte
+	var merkleLeaf [32]byte
+	copy(receiptRoot[:], beaconState.GetProcessedPowReceiptRootHash32())
```
Comment (nisdas) [inline beacon-chain/core/blocks/block_operations.go]: Same here

File: beacon-chain/core/blocks/block_operations.go [hunk 4]
```diff
@@ -524,6 +527,136 @@ func verifyAttestation(beaconState *pb.BeaconState, att *pb.Attestation) error {
...
+	); !ok {
+		return errors.New("Deposit merkle branch of PoW receipt root did not verify")
+	}
+
+	// We unmarshal the timestamp bytes into a time.Time value for us to use.
+	depositTimestampBytes := depositData[len(depositData)-8:]
+	depositUnixTime := int64(binary.BigEndian.Uint64(depositTimestampBytes))
+
+	// Parse beacon state's genesis time from a uint32 into a unix timestamp.
+	genesisUnixTime := int64(beaconState.GetGenesisTime())
```
Comment (nisdas) [inline beacon-chain/core/blocks/block_operations.go]: same here

File: beacon-chain/core/blocks/block_operations.go [hunk 5]
```diff
@@ -524,6 +527,136 @@ func verifyAttestation(beaconState *pb.BeaconState, att *pb.Attestation) error {
...
+	depositGenesisTimeDifference := depositUnixTime - genesisUnixTime
+	timeToLive := uint64(depositGenesisTimeDifference) / params.BeaconConfig().SlotDuration
+
+	// Verify current slot slot - allowed validator TTL is within the allowed boundary.
+	if beaconState.GetSlot()-timeToLive < params.BeaconConfig().ZeroBalanceValidatorTTL {
```
Comment (nisdas) [inline beacon-chain/core/blocks/block_operations.go]: same here

File: beacon-chain/core/state/state.go [hunk 6]
```diff
@@ -94,14 +96,22 @@ func InitialBeaconState(
...
+		depositValue := depositData[len(depositData)-16 : len(depositData)-8]
 		state, validatorIndex, err = v.ProcessDeposit(
 			state,
-			deposit.DepositData.DepositInput.Pubkey,
-			deposit.DepositData.Value,
-			deposit.DepositData.DepositInput.ProofOfPossession,
-			deposit.DepositData.DepositInput.WithdrawalCredentialsHash32,
-			deposit.DepositData.DepositInput.RandaoCommitmentHash32,
-			deposit.DepositData.DepositInput.PocCommitment,
+			depositInput.GetPubkey(),
```
Comment (nisdas) [inline beacon-chain/core/state/state.go]: here too

File: beacon-chain/core/blocks/block_operations.go [hunk 7]
```diff
@@ -524,6 +527,136 @@ func verifyAttestation(beaconState *pb.BeaconState, att *pb.Attestation) error {
...
+//     process_deposit(
+//       state=state,
+//       pubkey=deposit.deposit_data.deposit_input.pubkey,
+//       deposit=deposit.deposit_data.value,
+//       proof_of_possession=deposit.deposit_data.deposit_input.proof_of_possession,
+//       withdrawal_credentials=deposit.deposit_data.deposit_input.withdrawal_credentials,
+//       randao_commitment=deposit.deposit_data.deposit_input.randao_commitment,
+//       poc_commitment=deposit.deposit_data.deposit_input.poc_commitment,
+//     )
+func ProcessBlockValidatorDeposits(
```
Comment (terenc3t) [inline beacon-chain/core/blocks/block_operations.go]: Again, I'm favoring `blocks.ProcessValidatorDeposits` (optional)

File: beacon-chain/core/blocks/block_operations.go [hunk 8]
```diff
@@ -524,6 +527,136 @@ func verifyAttestation(beaconState *pb.BeaconState, att *pb.Attestation) error {
...
+		}
+		// depositData consists of depositInput []byte + depositValue [8]byte +
+		// depositTimestamp [8]byte.
+		depositValue := depositData[len(depositData)-16 : len(depositData)-8]
+		// We then mutate the beacon state with the verified validator deposit.
+		beaconState, _, err = v.ProcessDeposit(
+			beaconState,
+			depositInput.GetPubkey(),
+			binary.BigEndian.Uint64(depositValue),
+			depositInput.GetProofOfPossession(),
```
Comment (terenc3t) [inline beacon-chain/core/blocks/block_operations.go]: should we be using `GetX` here?

File: beacon-chain/core/blocks/block_operations.go [hunk 9]
```diff
@@ -524,6 +527,136 @@ func verifyAttestation(beaconState *pb.BeaconState, att *pb.Attestation) error {
...
+			depositInput.GetWithdrawalCredentialsHash32(),
+			depositInput.GetRandaoCommitmentHash32(),
+			depositInput.GetPocCommitment(),
+		)
+		if err != nil {
+			return nil, fmt.Errorf("could process deposit into beacon state: %v", err)
```
Comment (terenc3t) [inline beacon-chain/core/blocks/block_operations.go]: *could not

File: beacon-chain/core/blocks/block_operations.go [hunk 10]
```diff
@@ -524,6 +527,136 @@ func verifyAttestation(beaconState *pb.BeaconState, att *pb.Attestation) error {
...
+		}
+	}
+	return beaconState, nil
+}
+
+// DecodeDepositInput unmarshales a depositData byte slice into
+// a proto *pb.DepositInput by using the Simple Serialize (SSZ)
```
Comment (terenc3t) [inline beacon-chain/core/blocks/block_operations.go]: the depositData we get from lighthouse won't be in serialized proto format, but we can work interoperability with other client implementations later

File: beacon-chain/core/blocks/block_operations.go [hunk 11]
```diff
@@ -524,6 +527,136 @@ func verifyAttestation(beaconState *pb.BeaconState, att *pb.Attestation) error {
...
+	var merkleLeaf [32]byte
+	copy(receiptRoot[:], beaconState.GetProcessedPowReceiptRootHash32())
+	copy(merkleLeaf[:], depositData)
+	if ok := trie.VerifyMerkleBranch(
+		merkleLeaf,
+		deposit.GetMerkleBranchHash32S(),
+		params.BeaconConfig().DepositContractTreeDepth,
+		receiptRoot,
+	); !ok {
+		return errors.New("Deposit merkle branch of PoW receipt root did not verify")
```
Comment (terenc3t) [inline beacon-chain/core/blocks/block_operations.go]: we should probably print out the `receiptRoot` here, later debugging can be a pain

**Comment by rauljordan** [comment]
PTAL @nisdas

File: beacon-chain/core/blocks/block.go [hunk 12]
```diff
@@ -131,3 +134,30 @@ func DomainVersion(data *pb.ForkData, slot uint64, domainType uint64) uint64 {
 	constant := uint64(math.Pow(2, 32))
 	return ForkVersion(data, slot)*constant + domainType
 }
+
+// EncodeBlockDepositData converts a deposit input proto into an a byte slice
+// of Simple Serialized deposit input followed by 8 bytes for a deposit value
+// and 8 bytes for a unix timestamp, all in BigEndian format.
+func EncodeBlockDepositData(
```
[the remark this answers is not in the document]
Comment (terenc3t) [reply beacon-chain/core/blocks/block.go]: nvm, i realized you use `b.Encode...`, which is ok then

File: beacon-chain/core/blocks/block_operations.go (hunk 7 above)
Comment (nisdas) [reply beacon-chain/core/blocks/block_operations.go]: +1

**Comment by rauljordan** [comment]
PTAL @terenc3t

**Pull request merged**